resotocore/resotocore/infra_apps/package_manager.py

```diff
@@ -0,0 +1,264 @@
+
+
+def config_id(name: InfraAppName) -> ConfigId:
+    return ConfigId(f"resoto/apps/{name}")
+
+
+class PackageManager:
+    def __init__(
+        self,
+        entity_db: PackageEntityDb,
+        config_handler: ConfigHandler,
+        repos_cache_directory: Path = Path.home() / ".cache" / "resoto-infra-apps",
+    ) -> None:
+        """
+        Package manager for infra apps.
+
+        Args:
+            entity_db: EntityDb for storing app manifests
+            config_handler: ConfigHandler for storing app configs
+            repos_cache_directory: Directory where the infra apps repos are cloned
+        """
+
+        self.entity_db = entity_db
+        self.config_handler = config_handler
+        self.update_lock: Optional[Lock] = None
+        self.repos_cache_directory: Path = repos_cache_directory
+        self.cleanup_task: Optional[asyncio.Task[None]] = None
+
+    async def start(self) -> None:
+        self.update_lock = asyncio.Lock()
+        self.repos_cache_directory.mkdir(parents=True, exist_ok=True)
+
+    async def stop(self) -> None:
+        if self.cleanup_task:
+            self.cleanup_task.cancel()
+            await self.cleanup_task
+
+    def list(self) -> AsyncGenerator[InfraAppName, None]:
+        return self.entity_db.keys()
+
+    async def info(self, name: InfraAppName) -> Optional[AppManifest]:
+        if package := await self.entity_db.get(name):
+            return package.manifest
+        return None
+
+    async def update(self, name: InfraAppName) -> Optional[AppManifest]:
+        if not self.update_lock:
+            raise RuntimeError("PackageManager not started")
+
+        async with self.update_lock:
+            if package := await self.entity_db.get(name):
+                new_manifest = await self._fetch_manifest(name, package.source)
+                if not new_manifest:
+                    logger.warning(f"Failed to fetch manifest for app {name}, skipping update")
+                    return None
+
+                await self._delete(package.manifest.name)
+                return await self._install_from_manifest(new_manifest, package.source)
+        return None
+
+    async def update_all(self) -> None:
+        if not self.update_lock:
+            raise RuntimeError("PackageManager not started")
+
+        async with self.update_lock:
+            packages: Dict[InstallationSource, List[AppManifest]] = defaultdict(list)
+            async for package in self.entity_db.all():
+                packages[package.source].append(package.manifest)
+
+            for source, manifests in packages.items():
+                if isinstance(source, FromGit):
+                    repo_dir = await self._get_latest_git_repo(source.git_url)
+                    if not repo_dir:
+                        logger.warning(f"Failed to fetch git repo {source.git_url}, skipping update")
+                        continue
```

Comment:
how would this be possible to install?

Reply:
it is not, that's why we skip the update of this app.